In [ ]:
import requests
import pandas as pd

In [10]:
## API Request
url = "https://electric-vehicle-data2.p.rapidapi.com/records"

querystring = {"limit":"500","orderBy":"dataListIndex_asc"}

headers = {
	"X-RapidAPI-Key": "cdf8b1cb59mshf343db793d63080p109824jsn8aeb3d1ed903",
	"X-RapidAPI-Host": "electric-vehicle-data2.p.rapidapi.com"
}

response = requests.get(url, headers=headers, params=querystring)


data = response.json()
vehicles = data['data']
processed_data = []


for v in vehicles:
    
        # Extract necessary fields, adjusting keys as per the actual data structure
        v_county = v.get('county', '')
        v_city = v.get('city', '')
        v_modelyear = v.get('modelYear', '')
        v_make = v.get('make', '')
        v_model = v.get('model', '')
        v_electricVehicleType = v.get('electricVehicleType', '')
        v_electricRange = v.get('electricRange', '')
        v_electricUtility = v.get('electricUtility', '')

        # Append the extracted information as a dictionary to our list
        processed_data.append({
            'county': v_county,
            'city': v_city,
            'modelYear': v_modelyear,
            'make': v_make,
            'model': v_model,
            'electricVehicleType': v_electricVehicleType,
            'electricRange': v_electricRange,
            'electricUtility': v_electricUtility
        })
   

# Convert the processed data into a DataFrame
df = pd.DataFrame(processed_data)

# Now `df` contains only the specified number of rows.
# You can now export this DataFrame to a CSV file
df.to_csv('electricVehicledata.csv', index=False)

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

In [3]:
import pandas as pd

# Step 2: Load the CSV file
df = pd.read_csv('electricVehiclesxxx.csv')

# Step 3: Check for duplicate rows
duplicates = df.duplicated()

# Step 4: Count the number of duplicate rows
num_duplicates = duplicates.sum()

print(num_duplicates)
print(df[duplicates])

2543
         county           city  modelYear       make     model  \
10         King       Kirkland       2018      TESLA   MODEL 3   
26         King       Kirkland       2018      TESLA   MODEL 3   
33         King        Seattle       2015     NISSAN      LEAF   
85         King      Sammamish       2019      TESLA   MODEL 3   
109        King        Seattle       2019     NISSAN      LEAF   
...         ...            ...        ...        ...       ...   
2995       King       Issaquah       2023        BMW        IX   
2996   Thurston        Olympia       2017       FIAT       500   
2997  Jefferson  Port Townsend       2023  CHEVROLET  BOLT EUV   
2998       King        Seattle       2021      TESLA   MODEL Y   
2999       King      Shoreline       2015     NISSAN      LEAF   

                 electricVehicleType  electricRange  \
10    Battery Electric Vehicle (BEV)            215   
26    Battery Electric Vehicle (BEV)            215   
33    Battery Electric Vehicle (BEV) 

In [4]:
df_cleaned = df.drop_duplicates()
df_cleaned.to_csv('cleaned_electricVehiclesTest.csv', index=False)

### Data Aggregation
***
Data Aggregation usually refers to the process of gathering data from multiple sources and summarizing it for analysis. It might involve combining similar data from different datasets to get a more comprehensive view or to perform statistical analysis on a larger scale.


Due to the API's constraint of limiting data extraction to 500 rows per request(same rows results in each request), we will augment the dataset obtained from the API by incorporating an additional dataset sourced from Kaggle that have the same API's fields. This  will enable us to conduct a more extensive and nuanced analysis by leveraging a larger dataset.

Kaggle data source:https://www.kaggle.com/datasets/utkarshx27/electric-vehicle-population-data



In [11]:
# read the kaggle's dataset
df = pd.read_csv('kaggle_ev_data.csv')
df2 = pd.read_csv('electricVehicledata.csv')

# View first 10 rows
display(df.head(10))

#drop the fields that are not compatible with the API's fields
df = df.drop(['VIN (1-10)', 'State','Postal Code','Clean Alternative Fuel Vehicle (CAFV) Eligibility','Base MSRP','Legislative District','DOL Vehicle ID','Vehicle Location','2020 Census Tract'], axis=1)
df = df.iloc[600:3500]
# View first 10 rows after dropping the fields
display(df.head(10))


# Concatenate the DataFrames
combined_df = pd.concat([df, df2], ignore_index=True)
combined_df.to_csv('combined_dataset.csv', index=False)




VIN (1-10)     county          city State  Postal Code  modelYear  \
0  5YJ3E1EB4L     Yakima        Yakima    WA      98908.0       2020   
1  5YJ3E1EA7K  San Diego     San Diego    CA      92101.0       2019   
2  7JRBR0FL9M       Lane        Eugene    OR      97404.0       2021   
3  5YJXCBE21K     Yakima        Yakima    WA      98908.0       2019   
4  5UXKT0C5XH  Snohomish       Bothell    WA      98021.0       2017   
5  1N4AZ0CP4F  Snohomish       Everett    WA      98201.0       2015   
6  5YJ3E1EBXJ     Kitsap       Poulsbo    WA      98370.0       2018   
7  WDC0G5EB0K     Yakima        Naches    WA      98937.0       2019   
8  1N4AZ0CP3D     Kitsap  Port Orchard    WA      98366.0       2013   
9  KNDCC3LD9K     Kitsap        Olalla    WA      98359.0       2019   

            make      model                     electricVehicleType  \
0          TESLA    MODEL 3          Battery Electric Vehicle (BEV)   
1          TESLA    MODEL 3          Battery Electric Vehicle (BEV)   
2          VOLVO        S60  Plug-in Hybrid Electric Vehicle (PHEV)   
3          TESLA    MODEL X          Battery Electric Vehicle (BEV)   
4            BMW         X5  Plug-in Hybrid Electric Vehicle (PHEV)   
5         NISSAN       LEAF          Battery Electric Vehicle (BEV)   
6          TESLA    MODEL 3          Battery Electric Vehicle (BEV)   
7  MERCEDES-BENZ  GLC-CLASS  Plug-in Hybrid Electric Vehicle (PHEV)   
8         NISSAN       LEAF          Battery Electric Vehicle (BEV)   
9            KIA       NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  electricRange  Base MSRP  \
0           Clean Alternative Fuel Vehicle Eligible            322          0   
1           Clean Alternative Fuel Vehicle Eligible            220          0   
2             Not eligible due to low battery range             22          0   
3           Clean Alternative Fuel Vehicle Eligible            289          0   
4             Not eligible due to low battery range             14          0   
5           Clean Alternative Fuel Vehicle Eligible             84          0   
6           Clean Alternative Fuel Vehicle Eligible            215          0   
7             Not eligible due to low battery range             10          0   
8           Clean Alternative Fuel Vehicle Eligible             75          0   
9             Not eligible due to low battery range             26          0   

   Legislative District  DOL Vehicle ID             Vehicle Location  \
0                  14.0       127175366  POINT (-120.56916 46.58514)   
1                   NaN       266614659  POINT (-117.16171 32.71568)   
2                   NaN       144502018  POINT (-123.12802 44.09573)   
3                  14.0       477039944  POINT (-120.56916 46.58514)   
4                   1.0       106314946   POINT (-122.18384 47.8031)   
5                  38.0       107901699  POINT (-122.20596 47.97659)   
6                  23.0       475036313  POINT (-122.64681 47.73689)   
7                  14.0       338148968   POINT (-120.69972 46.7309)   
8                  26.0       249239623  POINT (-122.63847 47.54103)   
9                  26.0       475609368  POINT (-122.54729 47.42602)   

          electricUtility  2020 Census Tract  
0              PACIFICORP       5.307700e+10  
1                     NaN       6.073005e+09  
2                     NaN       4.103900e+10  
3              PACIFICORP       5.307700e+10  
4  PUGET SOUND ENERGY INC       5.306105e+10  
5  PUGET SOUND ENERGY INC       5.306104e+10  
6  PUGET SOUND ENERGY INC       5.303594e+10  
7              PACIFICORP       5.307700e+10  
8  PUGET SOUND ENERGY INC       5.303509e+10  
9  PUGET SOUND ENERGY INC       5.303509e+10

county              city  modelYear       make       model  \
600  Snohomish          Lynnwood       2022      TESLA     MODEL Y   
601     Kitsap      Port Orchard       2021      TESLA     MODEL 3   
602     Pierce        Eatonville       2021     TOYOTA  RAV4 PRIME   
603       King  Lake Forest Park       2013  CHEVROLET        VOLT   
604       King          Kirkland       2023      TESLA     MODEL Y   
605       King          Kirkland       2018      TESLA     MODEL 3   
606      Mason         Hoodsport       2022      TESLA     MODEL Y   
607  Snohomish     Granite Falls       2021      TESLA     MODEL 3   
608       King          Issaquah       2013      TESLA     MODEL S   
609       King        North Bend       2018   CHRYSLER    PACIFICA   

                        electricVehicleType  electricRange  \
600          Battery Electric Vehicle (BEV)              0   
601          Battery Electric Vehicle (BEV)              0   
602  Plug-in Hybrid Electric Vehicle (PHEV)             42   
603  Plug-in Hybrid Electric Vehicle (PHEV)             38   
604          Battery Electric Vehicle (BEV)              0   
605          Battery Electric Vehicle (BEV)            215   
606          Battery Electric Vehicle (BEV)              0   
607          Battery Electric Vehicle (BEV)              0   
608          Battery Electric Vehicle (BEV)            208   
609  Plug-in Hybrid Electric Vehicle (PHEV)             33   

                                       electricUtility  
600                             PUGET SOUND ENERGY INC  
601                             PUGET SOUND ENERGY INC  
602  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...  
603      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
604      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
605      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
606  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...  
607                             PUGET SOUND ENERGY INC  
608      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
609      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)

county         city  modelYear      make           model  \
0       King      Seattle       2020   HYUNDAI            KONA   
1       King      Bothell       2022      JEEP  GRAND CHEROKEE   
2     Yakima       Yakima       2023      JEEP  GRAND CHEROKEE   
3       King     Kirkland       2018     TESLA         MODEL 3   
4   Thurston      Olympia       2018       BMW              I3   
5  Snohomish   Marysville       2020     TESLA         MODEL 3   
6       King         Kent       2017  CHRYSLER        PACIFICA   
7       King  Woodinville       2020     TESLA         MODEL Y   
8     Island   Coupeville       2018     TESLA         MODEL 3   
9       King     Bellevue       2023     TESLA         MODEL Y   

                      electricVehicleType  electricRange  \
0          Battery Electric Vehicle (BEV)            258   
1  Plug-in Hybrid Electric Vehicle (PHEV)             25   
2  Plug-in Hybrid Electric Vehicle (PHEV)             25   
3          Battery Electric Vehicle (BEV)            215   
4  Plug-in Hybrid Electric Vehicle (PHEV)             97   
5          Battery Electric Vehicle (BEV)            266   
6  Plug-in Hybrid Electric Vehicle (PHEV)             33   
7          Battery Electric Vehicle (BEV)            291   
8          Battery Electric Vehicle (BEV)            215   
9          Battery Electric Vehicle (BEV)              0   

                                 electricUtility  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)  
1  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
2                                     PACIFICORP  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
4                         PUGET SOUND ENERGY INC  
5                         PUGET SOUND ENERGY INC  
6  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
7  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)  
8                         PUGET SOUND ENERGY INC  
9  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)

### 📝 Data Preprocessing and cleaning
****

#### Take a first look at the data

In [89]:
data  =pd.read_csv('combined_dataset.csv')
# Viewing the first few rows
print(data.head())

      county              city  modelYear       make       model  \
0  Snohomish          Lynnwood       2022      TESLA     MODEL Y   
1     Kitsap      Port Orchard       2021      TESLA     MODEL 3   
2     Pierce        Eatonville       2021     TOYOTA  RAV4 PRIME   
3       King  Lake Forest Park       2013  CHEVROLET        VOLT   
4       King          Kirkland       2023      TESLA     MODEL Y   

                      electricVehicleType  electricRange  \
0          Battery Electric Vehicle (BEV)              0   
1          Battery Electric Vehicle (BEV)              0   
2  Plug-in Hybrid Electric Vehicle (PHEV)             42   
3  Plug-in Hybrid Electric Vehicle (PHEV)             38   
4          Battery Electric Vehicle (BEV)              0   

                                     electricUtility  
0                             PUGET SOUND ENERGY INC  
1                             PUGET SOUND ENERGY INC  
2  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...  
3      PUG

In [90]:
data.shape

(3357, 8)

###### There are 3354 rows and 8 columns in the dataset.

In [91]:
data.dtypes

county                 object
city                   object
modelYear               int64
make                   object
model                  object
electricVehicleType    object
electricRange           int64
electricUtility        object
dtype: object

###### There are 2 numerical attributes and 6 categorical attributes in the dataset, model years for vehicles is represented as integers ,because they refer to a specific calendar year and are not used in calculations we will change its type to category.

In [92]:
# changing modelYear data type
data['modelYear'] = data['modelYear'].astype('category')

In [93]:

# Summary statistics for numerical features
print(data.describe())

       electricRange
count    3357.000000
mean       83.916592
std       101.080454
min         0.000000
25%         0.000000
50%        30.000000
75%       200.000000
max       330.000000


In [94]:
## how many rows that has the electric rang = 0
print((data['electricRange']==0).sum())

1243


###### we have observed an unusual distribution of values in the 'Electric Range' column, with a large number of vehicles having a value of 0. This could potentially indicate missing or unknown values in the dataset.

###### To account for this uncertainty and still make use of the available data we will fill the missing or zero electric range values with average values based on the vehicle model and type. 

In [95]:
# Identify entries with electricRange as 0
zero_range_filter = data['electricRange'] == 0

# Calculate the mean electricRange for each model and electricVehicleType combination
average_ranges = data[~zero_range_filter].groupby(['model', 'electricVehicleType'])['electricRange'].mean()

# Function to apply the replacement
def replace_zero_with_average(row):
    if row['electricRange'] == 0:  # If electricRange is 0, try to replace it
        # Use .get((row['model'], row['electricVehicleType']), default_value) to avoid KeyErrors
        return average_ranges.get((row['model'], row['electricVehicleType']), row['electricRange'])
    else:
        return row['electricRange']

# Apply the function
data['electricRange'] = data.apply(replace_zero_with_average, axis=1)

print((data['electricRange']==0).sum())

298


##### after running the code we could still see there's still 298 vehicles that has electric range = 0,we only replaced 945 vehicles with their mean electric range based on the model and electric vehicle type ,this could be one of the two reasons
### one, some of the vehicles that has the same model and electric vehicle type all have electric range = 0.Two, the vehicle that have electric range = 0 could have a unique model and type.
###### the 298 vehicles that have electric range = 0 will be dropped

In [96]:
# filter data to include vehicles with electric rang != 0
data = data[data['electricRange'] != 0]
print((data['electricRange']==0).sum())

0


##### Handeling missing value 
****


In [97]:
# Checking for any missing values
print(data.isnull().sum())

county                 0
city                   0
modelYear              0
make                   0
model                  0
electricVehicleType    0
electricRange          0
electricUtility        3
dtype: int64


In [98]:
# Find the rows where any cell has a missing value
rows_with_missing_values = data[data.isnull().any(axis=1)]

# Display the rows with missing values
print(rows_with_missing_values)

        county         city modelYear    make    model  \
115     Sonoma   Santa Rosa      2018   HONDA  CLARITY   
719   Thurston      Olympia      2019     KIA     NIRO   
1200   Pulaski  Little Rock      2019  NISSAN     LEAF   

                         electricVehicleType  electricRange electricUtility  
115   Plug-in Hybrid Electric Vehicle (PHEV)           47.0             NaN  
719   Plug-in Hybrid Electric Vehicle (PHEV)           26.0             NaN  
1200          Battery Electric Vehicle (BEV)          150.0             NaN  


###### for the 'electricUtility' column by imputation the utility data is typically associated with geography, we impute missing utilities based on the most common utility for that city and county.
###### -in check_1 we try to see the electric utility where county = Sonoma and city = santa Rosa ,the result is only 1 row(row 115) the one that contains the Nan ,we drop that row since we can't impute it.

###### -for check_2 we see that all the vehicles that have county = Thurston and = city Olympia has the electric utility = PUGET SOUND ENERGY INC, so we will replace the Nan with PUGET SOUND ENERGY INC.

###### -the same thing happen with check_1 resulted in check_3 , the result is only 1 row(row 1200) the one that contains the Nan ,we drop that row since we can't impute it.




In [99]:
check_1 = data[(data['county'] == 'Sonoma') & (data['city'] == 'Santa Rosa') ]
print(check_1)

# drop the single row resulted 
condition = (data['county'] == 'Sonoma') & (data['city'] == 'Santa Rosa')

# Invert the condition to select rows that do NOT match the criteria
data = data[~condition]


     county        city modelYear   make    model  \
115  Sonoma  Santa Rosa      2018  HONDA  CLARITY   

                        electricVehicleType  electricRange electricUtility  
115  Plug-in Hybrid Electric Vehicle (PHEV)           47.0             NaN  


In [100]:
check_2 = data[(data['county'] == 'Thurston') & (data['city'] == 'Olympia') ]
print(check_2)

## replace the rows that have county=Thurston and city=Olympia with electricUtility = PUGET SOUND ENERGY INC 
data.loc[(data['county'] == 'Thurston') & (data['city'] == 'Olympia'), 'electricUtility'] = 'PUGET SOUND ENERGY INC'

        county     city modelYear      make     model  \
105   Thurston  Olympia      2023     TESLA   MODEL Y   
126   Thurston  Olympia      2019       KIA      NIRO   
275   Thurston  Olympia      2018  CHRYSLER  PACIFICA   
434   Thurston  Olympia      2019       KIA      NIRO   
559   Thurston  Olympia      2022     TESLA   MODEL Y   
...        ...      ...       ...       ...       ...   
3299  Thurston  Olympia      2023       KIA      NIRO   
3301  Thurston  Olympia      2021     TESLA   MODEL Y   
3313  Thurston  Olympia      2021     TESLA   MODEL 3   
3314  Thurston  Olympia      2013    NISSAN      LEAF   
3353  Thurston  Olympia      2017      FIAT       500   

                         electricVehicleType  electricRange  \
105           Battery Electric Vehicle (BEV)     291.000000   
126           Battery Electric Vehicle (BEV)     239.000000   
275   Plug-in Hybrid Electric Vehicle (PHEV)      33.000000   
434           Battery Electric Vehicle (BEV)     239.000000   


In [101]:
check_3 = data[(data['county'] == 'Pulaski') & (data['city'] == 'Little Rock') ]
print(check_3)
 
#drop the single row resulted 
condition = (data['county'] == 'Pulaski') & (data['city'] == 'Little Rock')

# Invert the condition to select rows that do NOT match the criteria
data = data[~condition]


       county         city modelYear    make model  \
1200  Pulaski  Little Rock      2019  NISSAN  LEAF   

                 electricVehicleType  electricRange electricUtility  
1200  Battery Electric Vehicle (BEV)          150.0             NaN  


##### Handeling duplicates
****
###### we have 1013 duplicates that need to removed 


In [105]:
# Checking for duplicate rows
print(data.duplicated().sum())

# romve duplicates
data = data.drop_duplicates()


0
0
